src/main/scala/com/databricks/spark/avro/AvroSaver.scala

```diff
@@ -0,0 +1,205 @@
+        val builder = SchemaBuilder.builder()
+        val valueSchema = convertTypeToAvro(valueType, builder, structName)
+
+        newFieldBuilder.map().values(valueSchema)
+
+      case StructType(_) =>
+        convertStructToAvro(dataType.asInstanceOf[StructType],
+          newFieldBuilder.record(structName))
+
+      case other => sys.error(s"Unexpected type $dataType")
+    }
+  }
+
+  private def convertStructToAvro[T](structType: StructType, schemaBuilder: RecordBuilder[T]): T = {
+    val fieldsAssembler: FieldAssembler[T] = schemaBuilder.fields()
+    structType.fields foreach { field =>
+      val newField = fieldsAssembler.name(field.name).`type`()
+
+      if (field.nullable) {
+        convertTypeToAvro(field.dataType, newField.optional(), field.name)
+      } else {
+        convertFieldTypeToAvro(field.dataType, newField, field.name).noDefault()
+      }
+    }
+    fieldsAssembler.endRecord()
+  }
+
+  private def convertValueToAvro(item: Any, dataType: DataType, structName: String): Any = {
+    if (item == null) {
+      // Since some schema fields might be nullable, item can be null. If so, just return it
+      // without further processing
+      return null
+    }
+
+    dataType match {
+      case ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType | StringType |
+           BinaryType | BooleanType =>
+        item
+
+      case DecimalType() =>
+        item.toString
+
+      case TimestampType =>
+        item.asInstanceOf[Timestamp].getTime
+
+      case ArrayType(elementType, _) =>
+        val sourceArray = item.asInstanceOf[Seq[Any]]
+        val targetArray = new Array[Any](sourceArray.size)
+        var idx = 0
+
+        while (idx < sourceArray.size) {
+          targetArray(idx) = convertValueToAvro(sourceArray(idx), elementType, structName)
+          idx += 1
+        }
+
+        targetArray
+
+      case MapType(StringType, valueType, _) =>
+        val javaMap = new HashMap[String, Any]()
+        item.asInstanceOf[Map[String, Any]].foreach { case (key, value) =>
+          javaMap.put(key, convertValueToAvro(value, valueType, structName))
+        }
+        javaMap
+
+      case StructType(_) => rowToAvro(item.asInstanceOf[Row], dataType.asInstanceOf[StructType],
+        structName)._1.datum()
+    }
+  }
+
+  private def rowToAvro(row: Row, structType: StructType,
+      structName: String):(AvroKey[GenericRecord], NullWritable) = {
+    val builder = SchemaBuilder.record(structName)
+    val schema: Schema = convertStructToAvro(structType, builder)
+    val record = new Record(schema)
+
+    val fieldIterator = structType.fields.iterator
+    val rowIterator = row.iterator
+
+    while (fieldIterator.hasNext) {
+      val field = fieldIterator.next
+      record.put(field.name, convertValueToAvro(rowIterator.next, field.dataType, field.name))
+    }
+
+    (new AvroKey(record), NullWritable.get())
+  }
+
+  def save(schemaRDD: SchemaRDD, location: String): Unit = {
```

Comment:
This is a personal preference, but I like a source file to be laid out such that it can be read from top to bottom, as much as is feasible. I think you have basically done this, but inverted. If you prefer this style, that's fine, but if you have no preference, I think it'd be nice to reorder them such that a reader can start at this method and then basically just go down the file.